### ISSUE TITLE
Port the WsFederation middleware from Katana

### ISSUE THREAD

**Issue by Tratcher** [aspnet/Security#43]
Depends on System.IdentityModel.Tokens.Jwt and Microsoft.IdentityModel.Protocol.Extensions.

**Comment by xytyx** [comment]
Any news on WsFederation in ASP.NET 5 ?

**Comment by blowdart** [comment]
The dependency still stands and that work has not yet been completed by the team that owns the IdentityModel pieces.

**Comment by brockallen** [comment]
Any plans to add signoutcleanup support? That was missing in Katata v3's implementation. It's nice to have single sign-on, but single sign-out is also nice :)

**Comment by brentschmaltz** [comment]
signoutcleanup is on the map.

WsFed in on the map also, no time frame yet.

**Comment by guigobavaresco** [comment]
Hi, 
how about it, now?

**Comment by blowdart** [comment]
We're still awaiting support from the WAAD team, they're the ones driving it.

**Comment by brentschmaltz** [comment]
Yep, it is on our plates and radar. No promises till we drop OIDC.

**Comment by lomithrani** [comment]
Is there any alternative to authentify using wsfederation or salm2 meanwhile ?

**Comment by aredfox** [comment]
Is there any news on the feature / roadmap for this feature, as this is now holding back starting new applications in ASPNET5 for our organisation.

**Comment by marc-mueller** [comment]
Is there any update so far? Not being able to support SAML within our ASP.NET 5 application is currently a showstopper for our project since we have some dependencies with existing systems.

**Comment by Tratcher** [comment]
This is not happening for v1.0. We'll follow up afterwards.

**Comment by marc-mueller** [comment]
So this means that the development has not started yet? Is there any estimation about the release?

**Comment by Tratcher** [comment]
Correct. No, we haven't scheduled any of the post-v1 work yet.

**Comment by rschiefer** [comment]
We use ADFS extensively for 20 or so internal web applications.  This would be a major blocker for us to migrate to ASP.NET Core.

Can someone post a link to the related dependencies so we can go show support for that work as well?

**Comment by helmsb** [comment]
We've used ADFS as the core of our authentication for our internal application framework which is used throughout our organization. This is a huge blocker for us going to .NET Core.

**Comment by Tratcher** [comment]
@blowdart @brentschmaltz is it this one? https://github.com/dotnet/corefx/issues/4278

**Comment by MaximRouiller** [comment]
Just bouncing this again. I'm still seeing people trying the Katana bits with .NET Core. 

I'll refer them to this issue.

**Comment by leastprivilege** [comment]
Maybe it is just me - but I see a lot of companies using ADFS via WS-Fed. They are all blocked to move forward to ASP.NET Core (let alone .NET Core). 

Is this a way to push adoption of Windows Server 2016 ;) (I am afraid this does not work that way)

**Comment by marc-mueller** [comment]
I fully agree with @leastprivilege. There are so many large companies with their ADFS via WS-Fed setup and they won't change that fast. On the development side we are faster then on the infrastructure side and this would allow us to push ASP.NET Core.

**Comment by brentschmaltz** [comment]
@leastprivilege @marc-mueller @MaximRouiller @helmsb the roadmap to make this happen is fully understood. Everytime I bring it up, the beancounters ask who really cares, if you do contact your contacts here at MSFT directly.

@Tratcher dotnet/corefx#4278 is the start of it. CoreFx is the rightfull owner of SignedXml. Once that is in place, IM can re-introduce EnveopledSignatureReader (which should be in IM) and an updated SamlToken / Handler and WsFedSupport.

**Comment by MaximRouiller** [comment]
@brentschmaltz Right now, I don't. I just see a confusion around the packages.

I'll talk to my beancounter if I ever need it urgently.

**Comment by brentschmaltz** [comment]
@MaximRouiller I was referring my Microsoft beancounters who want to ensure the number 1 priorities are what we are focused on. So they need to hear from you.

**Comment by karelz** [comment]
Just to close the loop and to clarify: https://github.com/dotnet/corefx/issues/4278 is making progress towards being .NET Standard 2.0 extension, **however it is not the dependency blocking WS-Fed**.
More technical details are posted here: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/500#issuecomment-275218749.

**Comment by danroth27** [comment]
For folks looking for WS-Fed support in ASP.NET Core could you please chime in with:
1. What identity providers do you need to integrate with that require WS-Fed? We've heard from folks who want to use WS-Fed with older ADFS installations, but are there other identity providers you need to integrate with?
2. For folks wanting to use WS-Fed with ADFS is updating to Windows Server 2016 and using OIDC instead an option?
3. Do you need WS-Fed support on .NET Core (ex to run cross-platform) or would it be sufficient if this only worked on the full .NET Framework on Windows?

The more data we can get on your requirements the better. Thanks!

**Comment by onovotny** [comment]
I need this to be an IdP to SharePoint IaaS as it only supports WS-Fed 1.1.

**Comment by brockallen** [comment]
> For folks wanting to use WS-Fed with ADFS is updating to Windows Server 2016 and using OIDC instead an option?

This is not an option for most of my customers. If they could then they'd just use the OIDC support in 2016.

> Do you need WS-Fed support on .NET Core (ex to run cross-platform) or would it be sufficient if this only worked on the full .NET Framework on Windows

For now the full .NET framework would be sufficient, but I'd argue that eventually .NET Core will also be desirable.

**Comment by vsimonian** [comment]
>What identity providers do you need to integrate with that require WS-Fed? We've heard from folks who want to use WS-Fed with older ADFS installations, but are there other identity providers you need to integrate with?

At least in my use case, it's really just older ADFS installations.

>For folks wanting to use WS-Fed with ADFS is updating to Windows Server 2016 and using OIDC instead an option?

This is not an option for many reasons, be they bureaucratic or technical.

>Do you need WS-Fed support on .NET Core (ex to run cross-platform) or would it be sufficient if this only worked on the full .NET Framework on Windows?

Absolutely would need it on .NET Core - the application server is running RHEL.

**Comment by danroth27** [comment]
@vibronet

**Comment by marc-mueller** [comment]
Our main scenario is also to support older versions of ADFS which cannot be upgraded in the near future.

Concerning the target frameworks: Our target is to use the library cross platform, so .NET framework is not an option in our projects.

**Comment by leastprivilege** [comment]
> I need this to be an IdP to SharePoint IaaS

We are talking about the WS-Fed MW - not IdP functionality.

But anyways - WS-Fed IdP support for IdentityServer4 is coming soon.

**Comment by leastprivilege** [comment]
FTR

It is about older ADFS/ADFS proxies which cannot be easily upgraded (after all that's what you get for hard coupling ADFS with the Windows Server version).

While .NET Core support would be nice - I think full .NET is good enough for now and would be an easy port.

**Comment by onovotny** [comment]
@leastprivilege good to know -- misread the context then, thought it was for Ws-Fed all-up for the underlying signed xml libraries.

**Comment by Compufreak345** [comment]
I want to sign this as well - we need the support for WS-Fed because we need to use a customers old version of ADFS - upgrading to Windows Server 2016 is not an option. Full .NET would be completely OK for us, we already need to use it for other features.

**Comment by poke** [comment]
Adding my voice here as well. WS-Fed support on full .NET Framework would be enough for us.

Since this keeps getting mentioned, why is everyone referring to “*old* versions” of ADFS? We are using a *current* version of ADFS running on Windows Server 2016. Is there a way we can integrate that into ASP.NET Core? (We need ADFS for compatibility with certain legacy applications)

**Comment by danroth27** [comment]
@poke With Windows Server 2016 you should be able to use OpenID Connect instead of WS-Fed.

**Comment by poke** [comment]
@danroth27 Do you happen to have any example on how to configure that properly?

**Comment by danroth27** [comment]
Does this article help?: https://technet.microsoft.com/en-us/windows-server-docs/identity/ad-fs/development/enabling-openid-connect-with-ad-fs-2016

**Comment by poke** [comment]
@danroth27 Took me a while to transfer that over to ASP.NET Core, but yes, that article did help me, especially on the ADFS side. I’ve got it working now, thanks a lot!

**Comment by klings** [comment]
What @leastprivilege and @brockallen said. WS-Fed MW would give us a path to ASP.NET Core. ADFS is part of the problem.

**Comment by chrisdrobison** [comment]
For those of you who are interested, I ported the Katana WsFederation middleware over to ASP.NET Core. It has a hard dependency on the full .NET Framework. I have yet to build the Nuget package and publish it, but I'd love for anyone who wants to give it a go to see if it will be sufficient until something is put in place officially by Microsoft.

https://github.com/chrisdrobison/aspnetcore-wsfed

**Comment by Tratcher** [comment]
Nice work @chrisdrobison. We have something in the works that will use the latest version of IdentityModel to avoid dependency version conflicts. It will similarly be restricted to the full .NET Framework for now.

Side note: please don't use `Microsoft` in your package names if you intend to publish it (see https://github.com/aspnet-contrib).

**Comment by chrisdrobison** [comment]
@Tratcher Sounds good, I'll remove the name.

**Comment by leastprivilege** [comment]
@chrisdrobison nice! I have some real world usage for that. I can help you with review, testing and automated builds.

**Comment by chrisdrobison** [comment]
@leastprivilege Thanks! That would be much appreciated.

**Comment by Compufreak345** [comment]
@chrisdrobison 
Thanks for your work as well, I (probably misconfigured something and) created your first issue ;)

**Comment by leastprivilege** [comment]
@Tratcher do you have an ETA? the version conflict is icky.

**Comment by Tratcher** [comment]
Not quick, we're just beginning the work at the IM layer.

**Comment by chrisdrobison** [comment]
Do you happen to know if there is going to be a package newer than this published?

https://www.nuget.org/packages/Microsoft.IdentityModel.Protocols.WsFederation/2.0.0-beta8-305061149

I could attempt to upgrade the work I've done, but I fear all the other packages that one depends on are much newer now and this one has kind of been left behind.

**Comment by Tratcher** [comment]
Yes, that package will get updated/replaced.

**Comment by chrisdrobison** [comment]
Do you know when that will happen?

**Comment by Tratcher** [comment]
https://github.com/aspnet/Security/issues/43#issuecomment-288440812

**Comment by chrisdrobison** [comment]
I guess I could ILMerge the dependencies to get rid of the version conflict while you guys are working on that.

**Comment by Tratcher** [comment]
That's problematic as they are exposed as public API

**Comment by brentschmaltz** [comment]
@chrisdrobison current POR is for an April refresh of M.IM.Protocols.WsFederation. Currently this topic branch is used to piece it together.
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/tree/WsFed

**Comment by chrisdrobison** [comment]
@brentschmaltz I literally just ran into that right before you posted. Looking forward to that .

**Comment by DaleMckeown** [comment]
This is a blocking issue for us too. We need to be able to use WS-Fed as our ADFS server is running on 2012 r2 and cannot be upgraded without causing widespread issues.

**Comment by brentschmaltz** [comment]
@DaleMckeown we committed an early beta to https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet so progress is being made in this area.

**Comment by muratg** [comment]
@Eilon This needs to be assigned in our weekly triage meeting.

**Comment by DaleMckeown** [comment]
We managed to dump WS-Fed and implement oAuth 2 instead.

**Comment by abezulski** [comment]
@DaleMckeown Are you using ADFS3? If so, could you provide some samples? I'm porting MVC5 apps to .net core and having issues with ADFS integration.

**Comment by DaleMckeown** [comment]
@abezulski Yeah, we are using Oauth2 with ADFS3.0.

Can't give you a working example unfortunately, but me and @lilpug followed this post on carbon60: http://www.carbon60.com/blog/using-adfs-3-0-with-mvc-6-asp-net-5

Took us a while to modify the source code accordingly, but we go there in the end. David documented the changes made to get it working in a comment on the above blog post.

**Comment by YaoNiGu** [comment]
has anyone known any information about using  adfs 2.0 in core2.0 ?

**Comment by Tratcher** [comment]
@YaoNiGu It's not yet available. We've just started the work.

**Comment by ctro** [comment]
We need to access a legacy SOAP service that authenticates with WS-Security (WSS) from a .NET Core app.   We are looking for WSFed support also.
I work in Government IT.  .NET Core and Azure have much promise here, but we very often need to access legacy datasources.

**Comment by ashgadala** [comment]
Do you have an ETA on when this will this will be addressed.

**Comment by danroth27** [comment]
@ashgadala The work has started. I expect we'll have something in preview before the end of the year.

**Comment by ashgadala** [comment]
@danroth27  Wonderful. Thank you so much for the update.

**Comment by Tratcher** [comment]
The eager can start tracking the work in this branch: https://github.com/aspnet/security/tree/tratcher/wsfed.

The plan is to start by releasing a 2.0.0 compatible preview package off cycle from the normal ASP.NET Core milestones (e.g. ASAP). We'll let you know when something is available on myget.org.

**Comment by Tratcher** [comment]
Reminder - Apply this change when moving to 2.1: https://github.com/aspnet/Security/issues/1188

**Comment by Infoseeker** [comment]
@chrisdrobison - any updates on https://github.com/chrisdrobison/aspnetcore-wsfed - working with asp.net cpre 2 ?

**Comment by Tratcher** [comment]
An official preview is now available. See: https://github.com/aspnet/Security/issues/1473

**Comment by Tratcher** [comment]
Doc bug: https://github.com/aspnet/Docs/issues/4523

**Comment by Eilon** [comment]
Closing because the bulk of the work is done and in the public preview. We'll track any additional features and bugs as separate issues.
